File: solutions/022.generate-parentheses/generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
const generateParenthesis = n => {
  const res = []

  const bfs = (temp, left, right) => {
    if (temp.length === n * 2) {
      res.push(temp)
      return
    }

    if (left < n) {
      bfs(temp + '(', left + 1, right)
    }
    if (right < left) {
      bfs(temp + ')', left, right + 1)
    }
  }

  bfs('', 0, 0)

  return res
};

File: solutions/022.generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ret = []

        def compose(temp, left, right):
            nonlocal ret
            if left > right:
                return

            if left < 0:
                return
            if right < 0:
                return

            if not left and not right:
                ret.append(temp)
                return

            compose(temp + '(', left - 1, right)
            compose(temp + ')', left, right - 1)

        compose('', n, n)
        return ret
